src/Microsoft.Framework.Runtime.Hosting/Json/JsonDeserializer.cs

```diff
@@ -0,0 +1,167 @@
+
+        public JsonValue Deserialize(TextReader reader)
+        {
+            if (reader == null)
+            {
+                throw new ArgumentNullException(nameof(reader));
+            }
+
+            _buffer = new JsonBuffer(reader);
+            return Deserialize();
+        }
+
+        private JsonValue Deserialize()
+        {
+            var result = DeserializeInternal(_buffer.Read());
+
+            // There are still unprocessed char. The parsing is not finished. Error happened.
+            var nextToken = _buffer.Read();
+            if (nextToken.Type != JsonTokenType.EOL)
+            {
+                throw new JsonDeserializerException("Failed to continue deserializing. Additional token " + nextToken.Value, nextToken);
+            }
+
+            return result;
+        }
+
+        private JsonValue DeserializeInternal(JsonToken next)
+        {
+            if (next.Type == JsonTokenType.EOL)
+            {
+                return null;
+            }
+
+            if (next.Type == JsonTokenType.LeftSquareBracket)
+            {
+                return DeserializeArray(next);
+            }
+
+            if (next.Type == JsonTokenType.LeftCurlyBracket)
+            {
+                return DeserializeObject(next);
+            }
+
+            if (next.Type == JsonTokenType.String)
+            {
+                return new JsonString(next.Value, next.GetPosition());
+            }
+
+            if (next.Type == JsonTokenType.True || next.Type == JsonTokenType.False)
+            {
+                return new JsonBoolean(next);
+            }
+
+            if (next.Type == JsonTokenType.Null)
+            {
+                return new JsonNull(next.GetPosition());
+            }
+
+            if (next.Type == JsonTokenType.Number)
+            {
+                return new JsonNumber(next);
+            }
+
+            throw new JsonDeserializerException(JsonDeserializerResource.Format_UnexpectedToken(next.Value, next.Type), next);
+        }
+
+        private JsonArray DeserializeArray(JsonToken head)
+        {
+            var list = new List<JsonValue>();
+            while (true)
+            {
+                var next = _buffer.Read();
+                if (next.Type == JsonTokenType.RightSquareBracket)
+                {
+                    break;
+                }
+
+                list.Add(DeserializeInternal(next));
+
+                next = _buffer.Read();
+                if (next.Type == JsonTokenType.EOL)
+                {
+                    throw new JsonDeserializerException(JsonDeserializerResource.JSON_InvalidArrayEnd, next);
+                }
+                else if (next.Type == JsonTokenType.RightSquareBracket)
+                {
+                    break;
+                }
+                else if (next.Type != JsonTokenType.Comma)
+                {
+                    throw new JsonDeserializerException(JsonDeserializerResource.JSON_InvalidArrayExpectComma, next);
```

Comment:
Rather than having different message for each expected character, you can have a generic one like `Unexpected character. Expected: {0}`